### ISSUE TITLE
Task list: support raft learner in etcd.

### ISSUE THREAD

**Issue by jingyih** [etcd-io/etcd#10537]
***Background***
- [Design doc for raft learner (non-voting member)](https://github.com/etcd-io/etcd/blob/master/docs/server-learner.rst)
- Original issue #9161

***Open questions***
- [x] [~~How to evaluate learner progress~~](https://github.com/jingyih/etcd/issues/15)
- [x] [~~Leader does not respond to ReadIndex message from learner~~](https://github.com/jingyih/etcd/issues/7)

***Task List***
The list is subject to change.

**Phase I**
- [x] Create new feature branch - The feature branch will be merged back to master when the development is done. We will not squash commits to keep track of contribution.
- [x] API changes and regenerate go bindings.
   - add `isLearner` flag to `MemberAdd` API request.
   - add `MemberPromote` API.
   - add `isLearner` field in `Status` maintenance API response.

**Phase II**
Multiple developers can work in parallel on these tasks.
- [x] etcdserver and clientv3 support for `MemberAdd` with `isLearner` flag. Includes both routing the request to raft, and applying the result.
- [x] Add `etcdctl member add --learner`.
- [x] etcdserver and clientv3 implementation for `MemberPromote`.
- [x] Add `etcdctl member promote`.
- [x] Add `isLearner` field to `etcdctl member list`.
- [x] etcdserver and clientv3 support for `Status` maintenance API with `isLearner` field.
- [x] Add `isLearner` field to `etcdctl endpoint status`.
- [x] Learner reject all client requests except s-reads and status requests.
- [x] Exclude learner in leadership transfer.
- [x] Implement `canPromote()` check (before sending promote request to Raft consensus).
- [x] Add learner limit to etcdserver.
- [x] ~~Exclude learner nodes from clientv3 balancer endpoints.~~ -> clientv3 balancer retry when request send to learner fails.

**Phase III**
Multiple developers can work in parallel on these tasks.
- [x] Add metrics.
- [x] Integration test: add learner
- [x] Integration test: promote learner
- [x] Integration test: learner only accepts s-reads and status requests.
- [x] Integration test: cannot transfer leadership to learner
- [x] Integration test: add learner with large delta (CanPromote=false) requiring snapshot catchup
- [x] Integration test: clientv3 balancer with learner member
- [ ] Integration test: partitioned learner
- [ ] e2e test: `etcdctl member add --learner`, `etcdctl endpoint status`, `etcdctl member list`
- [ ] ~~Add auto-promote, but put behind a feature flag?~~ (moved to future work)
- [ ] ~~Throttle process of a learner catching up from leader.~~ (moved to future work)
- [x] Update etcdctl documentation
- [x] Update operations documentation
- [x] Update API documentation (auto generated when generating go bindings)
- [ ] ~~Update clientv3 documentation~~

**Future Work**
These items will likely not be included in v3.4 release.
- [ ] Add auto-promote, but put behind a feature flag?
- [ ] Throttle process of a learner catching up from leader.

***Pull Requests***
[Incremental PRs](https://github.com/jingyih/etcd/pulls?q=is%3Apr+is%3Aclosed) opened against feature branch.
[Aggregated PR](https://github.com/etcd-io/etcd/pull/10645/c) of feature branch against master branch.

**Comment by jingyih** [comment]
@WIZARD-CXY and @jingyih will be working on this. Please let us know if you want to help on some of the tasks.

**Comment by WIZARD-CXY** [comment]
LGTM, the task is clear. I can help in phase II and III.

**Comment by jingyih** [comment]
@WIZARD-CXY 

I created the feature branch here:
https://github.com/jingyih/etcd/tree/learner

For our development work, we can create pull request against jingyih/etcd/learner. In the end, we will merge all the changes from jingyih/etcd/learner to etcd-io/etcd/master. Since we will not squash commits, all the commits in feature branch will appear in master branch. Here is an example: https://github.com/etcd-io/etcd/pull/9860

Please let me know if you have any question.

**Comment by jpbetz** [comment]
Thanks Jingyi!

We did the same feature branch approach for the clientv3 balancer
improvements last summer. All author attribution of all commits is
retained, so it's works out really nice.

**Comment by WIZARD-CXY** [comment]
Roger that

**Comment by WIZARD-CXY** [comment]
@jingyih Maybe add `modify existing code to work with the learner implementation  ` to phase II.
Because I find that `TransferLeadership` function may transfer leader to a learner node.

**Comment by jingyih** [comment]
@WIZARD-CXY Thanks for pointing out:) Added. I changed the wording to be more specific: "Learner reject leadership transfer."

**Comment by WIZARD-CXY** [comment]
@jingyih maybe exclude the learner when transfer leadership. I think it more accurate in this case.

**Comment by jingyih** [comment]
> @jingyih maybe exclude the learner when transfer leadership. I think it is more accurate in this case.

Sounds good.

**Comment by purpleidea** [comment]
Can someone explain what this feature is? For anyone not working "face to face" internally on etcd, it's not obvious what this is. It was mentioned in the etcd community meeting, but I think everyone who attended was a paid RH employee. Thanks!

**Comment by WIZARD-CXY** [comment]
@purpleidea https://etcd.readthedocs.io/en/latest/server-learner.html

**Comment by xiang90** [comment]
>  For anyone not working "face to face" internally on etcd, it's not obvious what this is. It was mentioned in the etcd community meeting, but I think everyone who attended was a paid RH employee. Thanks!

This feature is currently driven by @jingyih from Google and @WIZARD-CXY from Alibaba. So it is a joint effort from the community :P. If you are interested, we definitely want your help.

**Comment by purpleidea** [comment]
@WIZARD-CXY Thanks for the links!

@xiang90 I didn't realize it wasn't just RH people. Will remove my comment, thanks for the info!

**Comment by purpleidea** [comment]
I just read

> https://etcd.readthedocs.io/en/latest/server-learner.html

I have some thoughts about this, based on my experience building etcd clusters automatically in https://github.com/purpleidea/mgmt/ If these insights are useful, I am happy to share. I don't have a lot of time to contribute new code for this particular implementation at the moment.

Some background:
* I first published this in 2016: https://purpleidea.com/blog/2016/06/20/automatic-clustering-in-mgmt/
* I re-wrote it from etcd V2 to V3.
* My golang skills are now improved significantly.

I am about one week away from releasing a re-write of this code to remove all the cruft caused by my lack of knowledge in golang.

Some thoughts:
* Instead of doing member add as a follower, wouldn't it be better for the new prospective member to connect as a client, and do some sort of "sync" to pull most of the data down, and *then* re-join as a member? This means existing cluster automation tools don't need to change much, they just need to add a single etcd client sync which is optional if they want the initial log data. It also means we DON'T need to break API for this.

* Currently there's a difficult race:
Suppose you have an N member cluster (perhaps N is 1 and hostname is A). You decide to add member B. You run member add API from A, and also member B starts up server. If there is any problem with startup, member A is in a borked state. What can you do to unbork? Raft log is blocked. A way to force remove the pending member would be useful.
If you did this operation with a "raft learner" the server is already started but not voting, so doing the switchover from non-voting server to voting server might be a more difficult problem to deal with, since it's already part of the cluster specification!

* I think the best way to model all this is to add MemberAPI operations into `etcd.Op` so that they can happen in a Txn. If this is something that intrigues anyone, then I can write up a proposal. A related proposal is: https://github.com/etcd-io/etcd/issues/10571 and https://github.com/etcd-io/etcd/issues/5277

Thanks for reading. If this is not helpful, please be blunt and tell me, and I won't make more noise.

**Comment by jingyih** [comment]
This task list was meant to be used for tracking the implementation progress. But I just added some context in the beginning. Sorry for the confusion.

**Comment by WIZARD-CXY** [comment]
@jingyih seems we are in the end of phase III, what else need I do?
```
Add metrics.
integration test: partitioned learner
e2e test: etcdctl member add --learner, etcdctl endpoint status, etcdctl member list
```

**Comment by WIZARD-CXY** [comment]
how about I work on the `add metrics`

**Comment by jingyih** [comment]
@WIZARD-CXY I am thinking about dropping `integration test: partitioned learner` and the e2e tests. We already added e2e test for `etcdctl member add --learner`, I do not feel like we need to add e2e test for `etcdctl endpoint status` and `etcdctl member list`. So probably the last one is `Add metrics`. Maybe first come up with a list of metrics that we want to add?

**Comment by jingyih** [comment]
We might want to add:

 - etcd_server_is_learner
Ref: https://github.com/etcd-io/etcd/blob/919b93b742c76b12a83bdf8885fa75f11db6bcac/docs/metrics/latest#L526

- We might also want some debugging metrics to count the number of total/failed/successful learner promotion.

**Comment by WIZARD-CXY** [comment]
Will do. I will submit a pr collecting the metrics

**Comment by jingyih** [comment]
> Will do. I will submit a pr collecting the metrics

Thanks. When you are ready, please create PR against etcd-io/master. Let's stop using the learner feature branch.

**Comment by WIZARD-CXY** [comment]
ok

**Comment by fabriziopandini** [comment]
@jingyih is it possible that an etcd member answer to request from clients while still "learning" see

**Comment by jingyih** [comment]
@fabriziopandini Are you testing with etcd binary built from the master branch? The learner feature is not released yet.

Learner does accept certain types of requests in the current implementation. I can provide more info if needed. Just want to first make sure if this is related to the issue you are facing.

**Comment by fabriziopandini** [comment]
@jingyih thanks for the quick answer.My use case is the following: 
I'm a kubeadm contributor and I'm investigating an issue on kind, where they are agreesively (in parallel) joining control-plane plane nodes to a K8s.

Joining control-plane nodes is a new kubeadm feature that creates a new etcd member on the joining node and then calls AddMember of the etcd V3 API (currently on etcd 3.1.10).

Now, the problem we are experiencing is that etcd / the API server starts to answer before the new etcd member is aligned with the rest of the nodes, giving false answers (after few seconds everything starts to behave as expected). From my understanding, what we are experiencing is etcd answering while still learning.

What I'm looking for is:
- possibly a confirmation that the above analysis makes sense (or at least that this is a possible condition)
- any suggestions about how to prevent this; the "learner" API is really promising, but I understand it still WIP and I'm open to any suggestion

Thank you in advance for your help

**Comment by jingyih** [comment]
> @jingyih thanks for the quick answer.My use case is the following:
> I'm a kubeadm contributor and I'm investigating an issue on kind, where they are agreesively (in parallel) joining control-plane plane nodes to a K8s.
> 
> Joining control-plane nodes is a new kubeadm feature that creates a new etcd member on the joining node and then calls AddMember of the etcd V3 API (currently on etcd 3.1.10).
> 
> Now, the problem we are experiencing is that etcd / the API server starts to answer before the new etcd member is aligned with the rest of the nodes, giving false answers (after few seconds everything starts to behave as expected). From my understanding, what we are experiencing is etcd answering while still learning.
> 
> What I'm looking for is:
> 
> * possibly a confirmation that the above analysis makes sense (or at least that this is a possible condition)
> * any suggestions about how to prevent this; the "learner" API is really promising, but I understand it still WIP and I'm open to any suggestion
> 
> Thank you in advance for your help

Please follow the instruction on how to add a new etcd member to cluster. The AddMember API should be called before starting the new etcd member, otherwise the existing cluster will not be able to recognize and verify the new member.

**Comment by arjunsingri** [comment]
How many learner nodes can an etcd cluster support? Does it make sense to replace watches with learner nodes? Can a learner remain a learner indefinitely?

**Comment by jingyih** [comment]
> How many learner nodes can an etcd cluster support? 

Currently in 3.4, the maximum number of learner is 1. This may increase in future.

> Does it make sense to replace watches with learner nodes? 

Currently learner does not support watch API. To scale watch you might want to use: https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/grpc_proxy.md#scalable-watch-api

> Can a learner remain a learner indefinitely?

Yes. Currently in 3.4 a leaner will not be auto promoted.
